### PR TITLE
Run backtracking assertion only in debug mode

### PR DESCRIPTION
**Pull request by elwayman02**
This wraps the `assertNotRendered` check in `runInDebug` so that it won't execute in production, even when either of the flags are enabled.

### CODE REVIEW

File: packages/ember-metal/lib/property_events.js [hunk 1]
```diff
@@ -113,8 +114,8 @@ function propertyDidChange(obj, keyName, _meta) {
   markObjectAsDirty(meta, keyName);
 
   if (isEnabled('ember-glimmer-detect-backtracking-rerender') ||
-      isEnabled('ember-glimmer-allow-backtracking-rerender')) {
-    assertNotRendered(obj, keyName, meta);
+    isEnabled('ember-glimmer-allow-backtracking-rerender')) {
+    runInDebug(() => { assertNotRendered(obj, keyName, meta) });
```
Comment (GavinJoyce) [inline packages/ember-metal/lib/property_events.js]: perhaps move the `runInDebug` to wrap the `if` statement too?

Comment (elwayman02) [inline packages/ember-metal/lib/property_events.js]: Seems reasonable

**Review by locks — APPROVED** [review]

**Review by Turbo87 — APPROVED** [review]

File: packages/ember-metal/lib/property_events.js [hunk 2]
```diff
@@ -112,10 +113,12 @@ function propertyDidChange(obj, keyName, _meta) {
 
   markObjectAsDirty(meta, keyName);
 
-  if (isEnabled('ember-glimmer-detect-backtracking-rerender') ||
+  runInDebug(() => {
+    if (isEnabled('ember-glimmer-detect-backtracking-rerender') ||
       isEnabled('ember-glimmer-allow-backtracking-rerender')) {
-    assertNotRendered(obj, keyName, meta);
-  }
+      assertNotRendered(obj, keyName, meta)
```
Comment (rwjblue) [inline packages/ember-metal/lib/property_events.js]: So I dug into this a bit, and it turns out that this isn't the right fix. Specifically, in the case where `ember-glimmer-allow-backtracking-rerender` is enabled we have to actually call `assertNotRendered` or we will not trigger a rerender when we hit this backtracking.  In other words, wrapping this in `runInDebug` does avoid an error (due to a bug in `assertNotRendered`) but it also breaks the ability for rerendering to happen.


I submitted https://github.com/emberjs/ember.js/pull/14876 to address.

**Review by rwjblue — CHANGES_REQUESTED** [review]

**Comment by rwjblue** [comment]
I'm going to close this in favor of the refactoring in https://github.com/emberjs/ember.js/pull/14876.  Thank you for pointing out the issue and taking a stab at fixing!

**Comment by vvohra** [comment]
@rwjblue we're trying to fix the issue in this ticket for 2.10.2. I see that your PR is against master and includes some other refactorings. It's hard to figure out if there's a simpler fix to get this working in 2.10.2.

**Comment by rwjblue** [comment]
Its likely that my commit in https://github.com/emberjs/ember.js/pull/14876 applies cleanly to release-2-10 branch.

**Comment by vvohra** [comment]
You're probably right, let me try it once more and run through tests.

**Pull request closed without merging**